dvc/objects/db/base.py

```diff
@@ -81,12 +81,16 @@ def add(
         fs: "BaseFileSystem",
         hash_info: "HashInfo",
         move: bool = True,
+        verify: Optional[bool] = None,
         **kwargs,
     ):
         if self.read_only:
             raise ObjectDBPermissionError("Cannot add to read-only ODB")
+
+        if verify is None:
+            verify = self.verify
         try:
-            self.check(hash_info, check_hash=self.verify)
+            self.check(hash_info, check_hash=verify)
             return
         except (ObjectFormatError, FileNotFoundError):
             pass
```

Comment:
Btw, discussed with @isidentical that for some filesystems like (hdfs and future ssh) `upload_fobj` down below will no longer be atomic, so we might need to use a temporary path here and then just `rename` into place. (there is an option to wrap fs calls to make them atomic but that is error prone and ugly).

Reply:
Seems to me like this might still need to be handled at the fs level, uploading to a temp path and renaming at the ODB level won't work for all of our filesystems (HTTP doesn't support move/rename)

Reply:
@pmrowla That atomicity is not something that fs should care about when uploading/downloading, this is an odb-level behavior.

> HTTP doesn't support move/rename

Are operations already atomic there? Or it just doesn't support rename at all anywhere?

Reply:
In the HTTP case, it's atomic since the full POST/PUT request wouldn't be completed so the server should drop whatever was partially uploaded. And yeah, we don't support rename/move/copy at all since there's no HTTP method for that operation (unless you're using an extension built on top of HTTP like webdav)

It seems to me that both `_upload` and `_upload_fobj` should work the same way, and should both guarantee atomicity at the fs level - like how in localfs we do the explicit upload to tempfile and rename for both _upload and _upload_fobj

Reply:
@pmrowla Thanks for clarifying! 

`_upload_fobj` is temporary until fsspec migration is complete and we can use `put/get[_file]` directly.

fs atomicity is unlikely to be guaranteed by all filesystems and might actually be undesirable in some use cases outside dvc (e.g. you might want to upload as much of a file as you can, or you might not care about atomicity so you might not want to waste an API call for `rename`), so it seems like it could be more robust if we do that in our odb layer (or fs wrapper after all?) for now.

Clearly, it seems like it would be useful to have the knowledge about whether or not particular fs operations are atomic so that we could waste the least api calls possible, so maybe our `fsspec_wrapper` is indeed a pretty good place for it for now, similar how, IIRC, in `C` libraries you have `atomic_*` functions. Maybe this could be useful for `fsspec` in general as well, not quite sure right now :thinking:

Reply:
Doesn't look like this PR is changing the old behaviour, so probably not worth blocking it because of it, but we'll def need to keep this in mind for the followups.